src/preamble.js

```diff
@@ -979,18 +980,20 @@ function createWasm() {
     exports = applySignatureConversions(exports);
 #endif
 
-    Module['asm'] = exports;
+    wasmExports = exports;
+    {{{ maybeExport('wasmExports') }}}
 
 #if PTHREADS
 #if MAIN_MODULE
-    registerTLSInit(Module['asm']['_emscripten_tls_init'], instance.exports, metadata);
+    registerTLSInit(wasmExports['_emscripten_tls_init'], instance.exports, metadata);
 #else
-    registerTLSInit(Module['asm']['_emscripten_tls_init']);
+    registerTLSInit(wasmExports['_emscripten_tls_init']);
 #endif
 #endif
 
 #if !IMPORTED_MEMORY
-    wasmMemory = Module['asm']['memory'];
+    wasmMemory = wasmExports['memory'];
+    {{{ maybeExport('wasmMemory') }}}
```

Comment:
Do we still need

https://github.com/emscripten-core/emscripten/blob/8b5c8687226a27e856f0449db5a92684c4614571/src/modules.js#L384

after this? I think that line can be removed as it doesn't work, as you found.

Reply:
I tried that, but without adding to the list is modules.js we don't get the `Aborted('wasmExports' was not exported. add it to EXPORTED_RUNTIME_METHODS)` messed when its not exported... and the test currently depends on that.

Reply:
I see. Can we perhaps move that logic to a shared place, and have `maybeExport` include it for that assertion? Can be a followup though.

Reply:
Perhaps... `maybeExports` is only needed for these very few items though.. where as the list in `modules.js` is that list of all possible runtime exports.   I'm not sure it makes sense to unify those two things.